rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala

```diff
@@ -0,0 +1,272 @@
+          if ! name.value.startsWith("_")
+        } yield {
+          <SECTION component="true" multivalued="true" name={name.value}>
+            <REPORTKEYS>
+              { for {
+                  call <- methodCalls
+                  method <- methods.get(call.methodId)
+                } yield {
+                  val value = call.parameters.get(method.classParameter).getOrElse("")
+                  <VALUE>{value}</VALUE>
+                }
+              }
+            </REPORTKEYS>
+          </SECTION>
+
+        }
+      }
+      </SECTIONS>
+    </TECHNIQUE>
+  }
+
+  def writeAll(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, commiter : EventActor) : Box[Unit] = {
+    for {
+      _ <- writeTechnique(technique, methods, modId, commiter)
+      res <- writeMetadata(technique, methods) ?~! "An error occured while writing metadata file"
+      _ <- archiver.commitFile(technique, metadataPath(technique), modId, commiter , s"Commiting Technique '${technique.bundleName.value}' metadata")
+    } yield {
+      res
+    }
+  }
+
+  def metadataPath(technique : Technique) = s"techniques/ncf_techniques/${technique.bundleName.value}/${technique.version.value}/metadata.xml"
+  def writeTechnique(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, commiter : EventActor) : Box[Unit] = {
+    import net.liftweb.util.Helpers.tryo
+    tryo {
+      for {
+        writer <- agentSpecific
+        file  <- writer.writeTechniqueFile(technique, methods)
+        _ <- file match {
+          case Some(file) => archiver.commitFile(technique, file, modId, commiter , s"Commiting Technique '${technique.bundleName.value}' file for agent ")
+          Full(file)
+          case None => Full("ok")
+        }
+      } yield {
+        file
+
+      }
+    }
+  }
+
+  def writeMetadata(technique : Technique, methods: Map[BundleName, GenericMethod]) : Box[Unit] = {
+    import net.liftweb.util.Helpers.tryo
+    tryo {
+      val content = techniqueMetadataContent(technique, methods)
+      val path = s"/var/rudder/configuration-repository/${metadataPath(technique)}"
+      val filePath = Paths.get(path)
+      if (!Files.exists(filePath))
+        Files.createFile(filePath)
+      scala.xml.XML.save(path, content)
+    }
+  }
+}
+
+trait AgentSpecificTechniqueWriter {
+
+  def writeTechniqueFile(technique : Technique, methods : Map[BundleName, GenericMethod] ) : Box[Option[String]]
+
+  def metadataContent(technique : Technique, methods : Map[BundleName, GenericMethod] ) : NodeSeq
+}
+
+class ClassicTechniqueWriter extends AgentSpecificTechniqueWriter {
+
+  def writeTechniqueFile(technique : Technique, methods : Map[BundleName, GenericMethod]) = Full(None)
+  def metadataContent(technique : Technique, methods : Map[BundleName, GenericMethod]) = {
+    val completeReporting = technique.methodCalls.exists(m => m.condition != "any" || m.condition != "cfengine-community" )
```

Comment:
this name implies that the reporting is complete, which is the opposite of what it needs: the reporting need to be completed.
what about "val extraReporting" ?

Reply:
or incompleteReporting